server/ladder_service/violation_service.py

```diff
@@ -0,0 +1,127 @@
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Optional
+
+import humanize
+
+from server.core import Service
+from server.decorators import with_logger
+from server.players import Player
+from server.timing import at_interval, datetime_now
+
+
+@dataclass
+class Violation():
+    count: int
+    time: datetime
+
+    def __init__(self, count: int = 1, time: Optional[datetime] = None):
+        self.count = count
+        self.time = time or datetime_now()
+
+    def register(self):
+        self.count += 1
+        self.time = datetime_now()
+
+    def get_ban_expiration(self) -> datetime:
+        if self.count < 2:
+            # No ban, expires as soon as it's registered
+            return self.time
+        elif self.count == 2:
+            return self.time + timedelta(minutes=10)
+        else:
+            return self.time + timedelta(minutes=30)
+
+    def get_remaining(self, now: Optional[datetime] = None) -> timedelta:
+        return self.get_ban_expiration() - (now or datetime_now())
+
+    def is_expired(self, now: Optional[datetime] = None) -> bool:
+        """
+        Whether the violation history should be reset. This is different from
+        the ban expiration time which should be checked by calling
+        `get_ban_expiration`.
+        """
+        now = now or datetime_now()
+        # TODO: Config?
+        return self.time + timedelta(hours=1) <= now
```

Comment:
Can we just make this an issue to fix later? Because it's gonna take a bit more refactoring effort than I want put in right now and I don't think it's worth blocking the PR for this.